File: src/services/circuit-breaker.ts
/**
 * RSS 熔断器 (Circuit Breaker)
 * 
 * 借鉴 WorldMonitor 的 per-feed 容错策略：
 * - 连续失败计数 → 自动进入冷却期
 * - 冷却期内返回缓存数据
 * - 成功后自动重置
 * - 内存缓存 + TTL 管理
 */

import { getLogger } from '../utils/logger';

const logger = getLogger('circuit-breaker');

export interface CircuitBreakerState {
  failureCount: number;
  cooldownUntil: number;
  lastSuccess: number;
}

export interface CacheEntry<T> {
  data: T;
  timestamp: number;
}

export class RSSCircuitBreaker {
  private states = new Map<string, CircuitBreakerState>();
  private cache = new Map<string, CacheEntry<any>>();

  // 配置参数
  private maxFailures: number;
  private cooldownMs: number;
  private cacheTtlMs: number;
  private maxCacheEntries: number;

  constructor(options?: {
    maxFailures?: number;     // 最大失败次数（默认 3）
    cooldownMs?: number;      // 冷却期毫秒（默认 5分钟）
    cacheTtlMs?: number;      // 缓存 TTL（默认 30分钟）
    maxCacheEntries?: number; // 最大缓存条目（默认 100）
  }) {
    this.maxFailures = options?.maxFailures ?? 3;
    this.cooldownMs = options?.cooldownMs ?? 5 * 60 * 1000;
    this.cacheTtlMs = options?.cacheTtlMs ?? 30 * 60 * 1000;
    this.maxCacheEntries = options?.maxCacheEntries ?? 100;
  }

  /**
   * 检查源是否处于冷却期
   */
  isOnCooldown(sourceKey: string): boolean {
    const state = this.states.get(sourceKey);
    if (!state) return false;
    if (Date.now() < state.cooldownUntil) return true;
    // 冷却期结束，重置
    if (state.cooldownUntil > 0) {
      this.states.delete(sourceKey);
    }
    return false;
  }

  /**
   * 记录抓取失败
   */
  recordFailure(sourceKey: string): void {
    const state = this.states.get(sourceKey) || { failureCount: 0, cooldownUntil: 0, lastSuccess: 0 };
    state.failureCount++;
    if (state.failureCount >= this.maxFailures) {
      state.cooldownUntil = Date.now() + this.cooldownMs;
      logger.warn(`[熔断] ${sourceKey} 连续失败 ${state.failureCount} 次，进入 ${this.cooldownMs / 1000}s 冷却期`);
    }
    this.states.set(sourceKey, state);
  }

  /**
   * 记录抓取成功（重置熔断状态）
   */
  recordSuccess(sourceKey: string): void {
    this.states.delete(sourceKey);
  }

  /**
   * 设置缓存
   */
  setCache<T>(key: string, data: T): void {
    this.cache.set(key, { data, timestamp: Date.now() });
    this.pruneCache();
  }

  /**
   * 获取缓存（如未过期）
   */
  getCache<T>(key: string): T | null {
    const entry = this.cache.get(key);
    if (!entry) return null;
    if (Date.now() - entry.timestamp > this.cacheTtlMs) {
      this.cache.delete(key);
      return null;
    }
    return entry.data as T;
  }

  /**
   * 获取过期缓存（作为 fallback，不检查 TTL）
   */
  getStaleCache<T>(key: string): T | null {
    const entry = this.cache.get(key);
    return entry ? (entry.data as T) : null;
  }

  /**
   * 清理过期缓存
   */
  private pruneCache(): void {
    if (this.cache.size <= this.maxCacheEntries) return;
    const entries = Array.from(this.cache.entries())
      .sort((a, b) => a[1].timestamp - b[1].timestamp);
    const toRemove = entries.slice(0, entries.length - this.maxCacheEntries);
    for (const [key] of toRemove) {
      this.cache.delete(key);
    }
  }

  /**
   * 获取所有熔断状态（用于监控/测试）
   */
  getStates(): Map<string, CircuitBreakerState> {
    return new Map(this.states);
  }

  /**
   * 获取统计摘要
   */
  getSummary(): { totalSources: number; onCooldown: number; cached: number } {
    const now = Date.now();
    let onCooldown = 0;
    for (const state of this.states.values()) {
      if (now < state.cooldownUntil) onCooldown++;
    }
    return {
      totalSources: this.states.size,
      onCooldown,
      cached: this.cache.size,
    };
  }
}

File: src/services/circuit_breaker.py
import time
from typing import Dict, Any, Optional
from ..utils.logger import get_logger

logger = get_logger("circuit-breaker")

class CircuitBreakerState:
    def __init__(self):
        self.failure_count = 0
        self.cooldown_until = 0
        self.last_success = 0

class CacheEntry:
    def __init__(self, data: Any):
        self.data = data
        self.timestamp = time.time()

class RSSCircuitBreaker:
    def __init__(self, max_failures: int = 3, cooldown_ms: int = 300000, cache_ttl_ms: int = 1800000, max_cache_entries: int = 100):
        self.max_failures = max_failures
        self.cooldown_ms = cooldown_ms
        self.cache_ttl_ms = cache_ttl_ms
        self.max_cache_entries = max_cache_entries
        self._states: Dict[str, CircuitBreakerState] = {}
        self._cache: Dict[str, CacheEntry] = {}

    def is_on_cooldown(self, source_key: str) -> bool:
        state = self._states.get(source_key)
        if not state: return False
        now = time.time() * 1000
        if now < state.cooldown_until: return True
        
        # 冷却期结束，重置
        if state.cooldown_until > 0:
            del self._states[source_key]
        return False

    def record_failure(self, source_key: str):
        state = self._states.get(source_key)
        if not state:
            state = CircuitBreakerState()
            self._states[source_key] = state
        
        state.failure_count += 1
        if state.failure_count >= self.max_failures:
            state.cooldown_until = (time.time() * 1000) + self.cooldown_ms
            logger.warning(f"[熔断] {source_key} 连续失败 {state.failure_count} 次，进入 {self.cooldown_ms / 1000}s 冷却期")

    def record_success(self, source_key: str):
        if source_key in self._states:
            del self._states[source_key]

    def set_cache(self, key: str, data: Any):
        self._cache[key] = CacheEntry(data)
        self._prune_cache()

    def get_cache(self, key: str) -> Optional[Any]:
        entry = self._cache.get(key)
        if not entry: return None
        if (time.time() - entry.timestamp) * 1000 > self.cache_ttl_ms:
            del self._cache[key]
            return None
        return entry.data

    def get_stale_cache(self, key: str) -> Optional[Any]:
        entry = self._cache.get(key)
        return entry.data if entry else None

    def _prune_cache(self):
        if len(self._cache) <= self.max_cache_entries: return
        # 按时间戳排序，移除最旧的
        sorted_keys = sorted(self._cache.keys(), key=lambda k: self._cache[k].timestamp)
        for k in sorted_keys[:len(self._cache) - self.max_cache_entries]:
            del self._cache[k]

    def get_summary(self) -> Dict[str, int]:
        now = time.time() * 1000
        on_cooldown = sum(1 for s in self._states.values() if now < s.cooldown_until)
        return {
            "total_sources": len(self._states),
            "on_cooldown": on_cooldown,
            "cached": len(self._cache)
        }
